Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -1443,6 +1448,160 @@ extension SourceKitServer {
+    languageService.symbolInfo(request)
+  }
+
+  /// Extracts our implementation-specific data about a type hierarchy
+  /// item as encoded in `indexToLSPTypeHierarchyItem`.
+  /// 
+  /// - Parameter data: The opaque data structure to extract
+  /// - Returns: The extracted data if successful or nil otherwise
+  private func extractTypeHierarchyItemData(_ rawData: LSPAny?) -> (uri: DocumentURI, usr: String)? {
+    guard case let .dictionary(data) = rawData,
+          case let .string(uriString) = data["uri"],
+          case let .string(usr) = data["usr"] else {
+      return nil
+    }
+    return (
+      uri: DocumentURI(string: uriString),
+      usr: usr
+    )
+  }
+
+  func supertypes(_ req: Request<TypeHierarchySupertypesRequest>) {
+    guard let data = extractTypeHierarchyItemData(req.params.item.data),
+          let index = self.workspaceForDocument(uri: data.uri)?.index else {
+      req.reply([])
+      return
+    }
+    let occurs = index.occurrences(relatedToUSR: data.usr, roles: .baseOf)
+    let types = occurs.compactMap { occurrence -> TypeHierarchyItem? in
+      guard let location = indexToLSPLocation(occurrence.location) else {
+        return nil
+      }
+
+      // Resolve the supertype's definition to find its location
+      let definition = index.occurrences(ofUSR: occurrence.symbol.usr, roles: [.definition, .declaration]).first
+      let definitionSymbolLocation = definition?.location
+      let definitionLocation = definitionSymbolLocation.flatMap(indexToLSPLocation)
+
+      return indexToLSPTypeHierarchyItem(
+        symbol: occurrence.symbol,
+        moduleName: definitionSymbolLocation?.moduleName,
+        location: definitionLocation ?? location, // Use occurrence location as fallback
+        index: index
+      )
+    }
+    req.reply(types)
+  }
+
+  func subtypes(_ req: Request<TypeHierarchySubtypesRequest>) {
+    guard let data = extractTypeHierarchyItemData(req.params.item.data),
+          let index = self.workspaceForDocument(uri: data.uri)?.index else {
+      req.reply([])
+      return
+    }
+    let occurs = index.occurrences(ofUSR: data.usr, roles: [.baseOf, .extendedBy])
+    let types = occurs.compactMap { occurrence -> TypeHierarchyItem? in
+      guard let location = indexToLSPLocation(occurrence.location),
+            let related = occurrence.relations.first else {
+        return nil
+      }
+
+      // Resolve the subtype's definition to find its location
+      let definition = index.occurrences(ofUSR: related.symbol.usr, roles: [.definition, .declaration]).first
+      let definitionSymbolLocation = definition.map(\.location)
+      let definitionLocation = definitionSymbolLocation.flatMap(indexToLSPLocation)
+
+      return indexToLSPTypeHierarchyItem(
+        symbol: related.symbol,
+        moduleName: definitionSymbolLocation?.moduleName,
+        location: definitionLocation ?? location, // Use occurrence location as fallback
+        index: index
+      )
```

Comment:
I wonder whether it would make sense to extract that code to a function because it’s shared between super type and subtype. 

And now that I think about it, we could probably also extract the closure in `incomingCalls`/`outgoingCalls`.

Reply:
Yeah that would be nice indeed. The main difficulty with call hierarchy is that `CallHierarchyIncomingCall` and `CallHierarchyOutgoingCall` are different types

Reply:
Did you mean type hierarchy? The problem is that there are still some subtleties to be considered, e.g. the `subtypes` closure uses the related symbol rather than the occurrence directly, leaving us essentially with only the definition lookup, which we could abstract over (also `subtypes` will get some more special logic with https://github.com/apple/sourcekit-lsp/pull/582/files#r916912420). Since we need both the definition and the symbol location, we'd need these values separately anyway. If you feel strongly about this, I could try and see, but I don't think we would gain much from factoring/unifying these methods.

Reply:
Yes, I meant type hierarchy.

Oh, I missed those differences. In that case I withdraw my commend and agree that we don’t need to unify these.